examples/allocation/rust/testdata/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+extern crate alloc;
+extern crate core;
+extern crate wee_alloc;
+
+use alloc::vec::Vec;
+use std::slice;
+use std::mem::MaybeUninit;
+
+/// Prints a greeting to the console using [`log`].
+fn greet(name: &String) {
+    log(&["wasm >> ", &greeting(name)].concat());
+}
+
+/// gets a greeting for the name.
+fn greeting(name: &String) -> String {
+    return ["Hello, ", &name, "!"].concat();
+}
+
+/// Logs a message to the console using [`_log`].
+fn log(message: &String) {
+    unsafe {
+        let (ptr, len) = string_to_ptr(message);
+        _log(ptr, len);
+    }
+}
+
+#[link(wasm_import_module = "env")]
+extern "C" {
+    /// WebAssembly import which prints a string (linear memory offset,
+    /// byteCount) to the console.
+    ///
+    /// Note: This is not an ownership transfer: Rust still owns the pointer
+    /// and ensures it isn't deallocated during this call.
+    #[link_name = "log"]
+    fn _log(ptr: u32, size: u32);
+}
+
+/// WebAssembly export that accepts a string (linear memory offset, byteCount)
+/// and calls [`greet`].
+///
+/// Note: The input parameters were returned by [`allocate`]. This is not an
+/// ownership transfer, so the inputs can be reused after this call.
+#[cfg_attr(
+all(target_arch = "wasm32", target_os = "unknown"),
+export_name = "greet"
+)]
+#[no_mangle]
+pub unsafe extern "C" fn _greet(ptr: u32, len: u32) {
```

Comment:
Note: I used direct wasm types here as I noticed rust will sometimes compile different arity. For example, if you return multiple values it switches them into out params in wasm. Having this defined the way it looks in wasm is helpful to me as it prevents odd signature changes.